File: lib/foundation/json.ml
type t =
  | Null
  | Bool of bool
  | Int of int
  | Int64 of int64
  | String of string
  | Array of t list
  | Object of (string * t) list

type error = { offset : int; message : string }

let escape_string value =
  let buffer = Buffer.create 64 in
  String.iter
    (fun character ->
      match character with
      | '"' -> Buffer.add_string buffer "\\\""
      | '\\' -> Buffer.add_string buffer "\\\\"
      | '\b' -> Buffer.add_string buffer "\\b"
      | '\012' -> Buffer.add_string buffer "\\f"
      | '\n' -> Buffer.add_string buffer "\\n"
      | '\r' -> Buffer.add_string buffer "\\r"
      | '\t' -> Buffer.add_string buffer "\\t"
      | character when Char.code character < 0x20 ->
          Buffer.add_string buffer
            (Printf.sprintf "\\u%04x" (Char.code character))
      | character -> Buffer.add_char buffer character)
    value;
  Buffer.contents buffer

let rec canonical_to_buffer buffer = function
  | Null -> Buffer.add_string buffer "null"
  | Bool true -> Buffer.add_string buffer "true"
  | Bool false -> Buffer.add_string buffer "false"
  | Int value -> Buffer.add_string buffer (string_of_int value)
  | Int64 value -> Buffer.add_string buffer (Int64.to_string value)
  | String value ->
      Buffer.add_char buffer '"';
      Buffer.add_string buffer (escape_string value);
      Buffer.add_char buffer '"'
  | Array values ->
      Buffer.add_char buffer '[';
      List.iteri
        (fun index value ->
          if index > 0 then Buffer.add_char buffer ',';
          canonical_to_buffer buffer value)
        values;
      Buffer.add_char buffer ']'
  | Object fields ->
      Buffer.add_char buffer '{';
      fields
      |> List.sort (fun (left, _) (right, _) -> String.compare left right)
      |> List.iteri (fun index (key, value) ->
          if index > 0 then Buffer.add_char buffer ',';
          canonical_to_buffer buffer (String key);
          Buffer.add_char buffer ':';
          canonical_to_buffer buffer value);
      Buffer.add_char buffer '}'

let to_string value =
  let buffer = Buffer.create 256 in
  canonical_to_buffer buffer value;
  Buffer.contents buffer

let to_pretty_string value =
  let buffer = Buffer.create 512 in
  let rec emit indent = function
    | (Null | Bool _ | Int _ | Int64 _ | String _) as scalar ->
        canonical_to_buffer buffer scalar
    | Array [] -> Buffer.add_string buffer "[]"
    | Array values ->
        Buffer.add_string buffer "[\n";
        List.iteri
          (fun index child ->
            Buffer.add_string buffer (String.make (indent + 2) ' ');
            emit (indent + 2) child;
            if index + 1 < List.length values then Buffer.add_char buffer ',';
            Buffer.add_char buffer '\n')
          values;
        Buffer.add_string buffer (String.make indent ' ');
        Buffer.add_char buffer ']'
    | Object [] -> Buffer.add_string buffer "{}"
    | Object fields ->
        let fields =
          List.sort
            (fun (left, _) (right, _) -> String.compare left right)
            fields
        in
        Buffer.add_string buffer "{\n";
        List.iteri
          (fun index (key, child) ->
            Buffer.add_string buffer (String.make (indent + 2) ' ');
            canonical_to_buffer buffer (String key);
            Buffer.add_string buffer ": ";
            emit (indent + 2) child;
            if index + 1 < List.length fields then Buffer.add_char buffer ',';
            Buffer.add_char buffer '\n')
          fields;
        Buffer.add_string buffer (String.make indent ' ');
        Buffer.add_char buffer '}'
  in
  emit 0 value;
  Buffer.add_char buffer '\n';
  Buffer.contents buffer

exception Parse_error of error

let parse source =
  let length = String.length source and offset = ref 0 in
  let fail message = raise (Parse_error { offset = !offset; message }) in
  let peek () = if !offset < length then Some source.[!offset] else None in
  let take () =
    match peek () with
    | None -> fail "unexpected end of JSON"
    | Some character ->
        incr offset;
        character
  in
  let rec whitespace () =
    match peek () with
    | Some (' ' | '\t' | '\r' | '\n') ->
        incr offset;
        whitespace ()
    | _ -> ()
  in
  let expect_literal literal value =
    String.iter
      (fun expected -> if take () <> expected then fail ("expected " ^ literal))
      literal;
    value
  in
  let hex_value = function
    | '0' .. '9' as character -> Char.code character - Char.code '0'
    | 'a' .. 'f' as character -> Char.code character - Char.code 'a' + 10
    | 'A' .. 'F' as character -> Char.code character - Char.code 'A' + 10
    | _ -> fail "invalid Unicode escape"
  in
  let add_utf8 buffer codepoint =
    if codepoint <= 0x7f then Buffer.add_char buffer (Char.chr codepoint)
    else if codepoint <= 0x7ff then (
      Buffer.add_char buffer (Char.chr (0xc0 lor (codepoint lsr 6)));
      Buffer.add_char buffer (Char.chr (0x80 lor (codepoint land 0x3f))))
    else if codepoint <= 0xffff then (
      Buffer.add_char buffer (Char.chr (0xe0 lor (codepoint lsr 12)));
      Buffer.add_char buffer (Char.chr (0x80 lor ((codepoint lsr 6) land 0x3f)));
      Buffer.add_char buffer (Char.chr (0x80 lor (codepoint land 0x3f))))
    else (
      Buffer.add_char buffer (Char.chr (0xf0 lor (codepoint lsr 18)));
      Buffer.add_char buffer
        (Char.chr (0x80 lor ((codepoint lsr 12) land 0x3f)));
      Buffer.add_char buffer (Char.chr (0x80 lor ((codepoint lsr 6) land 0x3f)));
      Buffer.add_char buffer (Char.chr (0x80 lor (codepoint land 0x3f))))
  in
  let unicode_escape () =
    let value = ref 0 in
    for _ = 1 to 4 do
      value := (!value lsl 4) lor hex_value (take ())
    done;
    !value
  in
  let parse_string () =
    if take () <> '"' then fail "expected string";
    let buffer = Buffer.create 32 in
    let rec loop () =
      match take () with
      | '"' -> Buffer.contents buffer
      | '\\' ->
          (match take () with
          | '"' -> Buffer.add_char buffer '"'
          | '\\' -> Buffer.add_char buffer '\\'
          | '/' -> Buffer.add_char buffer '/'
          | 'b' -> Buffer.add_char buffer '\b'
          | 'f' -> Buffer.add_char buffer '\012'
          | 'n' -> Buffer.add_char buffer '\n'
          | 'r' -> Buffer.add_char buffer '\r'
          | 't' -> Buffer.add_char buffer '\t'
          | 'u' ->
              let first = unicode_escape () in
              if first >= 0xd800 && first <= 0xdbff then (
                if take () <> '\\' || take () <> 'u' then
                  fail "high surrogate must be followed by a low surrogate";
                let second = unicode_escape () in
                if second < 0xdc00 || second > 0xdfff then
                  fail "high surrogate must be followed by a low surrogate";
                add_utf8 buffer
                  (0x10000 + ((first - 0xd800) lsl 10) + (second - 0xdc00)))
              else if first >= 0xdc00 && first <= 0xdfff then
                fail "lone low surrogate is invalid"
              else add_utf8 buffer first
          | _ -> fail "invalid string escape");
          loop ()
      | character when Char.code character < 0x20 ->
          fail "unescaped control character"
      | character ->
          Buffer.add_char buffer character;
          loop ()
    in
    loop ()
  in
  let parse_number () =
    let start = !offset in
    (match peek () with
    | Some '-' -> incr offset
    | _ -> ());
    (match peek () with
    | Some '0' -> (
        incr offset;
        match peek () with
        | Some '0' .. '9' -> fail "leading zero in JSON number"
        | _ -> ())
    | Some '1' .. '9' ->
        while
          match peek () with
          | Some '0' .. '9' ->
              incr offset;
              true
          | _ -> false
        do
          ()
        done
    | _ -> fail "invalid number");
    (match peek () with
    | Some ('.' | 'e' | 'E') -> fail "runner JSON permits integers only"
    | _ -> ());
    let raw = String.sub source start (!offset - start) in
    try
      let value = Int64.of_string raw in
      if value <= Int64.of_int max_int && value >= Int64.of_int min_int then
        Int (Int64.to_int value)
      else Int64 value
    with Failure _ -> fail "integer is out of range"
  in
  let rec value () =
    whitespace ();
    match peek () with
    | Some 'n' -> expect_literal "null" Null
    | Some 't' -> expect_literal "true" (Bool true)
    | Some 'f' -> expect_literal "false" (Bool false)
    | Some '"' -> String (parse_string ())
    | Some '[' -> array ()
    | Some '{' -> object_ ()
    | Some ('-' | '0' .. '9') -> parse_number ()
    | Some _ -> fail "expected a JSON value"
    | None -> fail "expected a JSON value"
  and array () =
    ignore (take ());
    whitespace ();
    if peek () = Some ']' then (
      ignore (take ());
      Array [])
    else
      let rec items accumulator =
        let child = value () in
        whitespace ();
        match take () with
        | ']' -> Array (List.rev (child :: accumulator))
        | ',' -> items (child :: accumulator)
        | _ -> fail "expected ',' or ']'"
      in
      items []
  and object_ () =
    ignore (take ());
    whitespace ();
    if peek () = Some '}' then (
      ignore (take ());
      Object [])
    else
      let rec fields seen accumulator =
        whitespace ();
        let key = parse_string () in
        if List.mem key seen then fail ("duplicate JSON object key: " ^ key);
        whitespace ();
        if take () <> ':' then fail "expected ':'";
        let child = value () in
        whitespace ();
        match take () with
        | '}' -> Object (List.rev ((key, child) :: accumulator))
        | ',' -> fields (key :: seen) ((key, child) :: accumulator)
        | _ -> fail "expected ',' or '}'"
      in
      fields [] []
  in
  if not (Util.valid_utf8 source) then
    Error { offset = 0; message = "JSON input is not valid UTF-8" }
  else
    try
      let parsed = value () in
      whitespace ();
      if !offset <> length then fail "trailing JSON input";
      Ok parsed
    with Parse_error error -> Error error

let member name = function
  | Object fields -> List.assoc_opt name fields
  | _ -> None

let as_string = function
  | String value -> Some value
  | _ -> None

let as_int = function
  | Int value -> Some value
  | _ -> None

let as_bool = function
  | Bool value -> Some value
  | _ -> None

let as_array = function
  | Array values -> Some values
  | _ -> None

let as_object = function
  | Object fields -> Some fields
  | _ -> None

let exact_object ~context ~allowed = function
  | Object fields -> (
      match
        List.find_opt (fun (name, _) -> not (List.mem name allowed)) fields
      with
      | Some (name, _) ->
          Error (Printf.sprintf "%s has unknown field %s" context name)
      | None -> Ok fields)
  | _ -> Error (context ^ " must be an object")

File: crates/foundation/src/json.rs
use std::collections::{BTreeMap, BTreeSet};
use std::fmt;

#[derive(Clone, Debug, Eq, PartialEq)]
pub enum JsonValue {
    Null,
    Boolean(bool),
    Integer(i64),
    String(String),
    Array(Vec<Self>),
    Object(BTreeMap<String, Self>),
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct JsonLimits {
    pub max_bytes: usize,
    pub max_depth: u32,
    pub max_values: usize,
}

impl Default for JsonLimits {
    fn default() -> Self {
        Self {
            max_bytes: 16 * 1024 * 1024,
            max_depth: 128,
            max_values: 1_000_000,
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct JsonError {
    pub offset: usize,
    pub message: String,
}

impl fmt::Display for JsonError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(formatter, "JSON byte {}: {}", self.offset, self.message)
    }
}

impl std::error::Error for JsonError {}

impl JsonValue {
    /// Parse language-independent product JSON.
    ///
    /// # Errors
    /// Rejects malformed syntax, duplicate keys, non-integer numbers, and
    /// values outside the default deterministic resource envelope.
    pub fn parse(source: &str) -> Result<Self, JsonError> {
        Self::parse_with_limits(source, JsonLimits::default())
    }

    /// Parse product JSON from an explicitly validated UTF-8 byte boundary.
    ///
    /// # Errors
    /// Returns the first UTF-8 or strict JSON contract violation.
    pub fn parse_bytes(source: &[u8]) -> Result<Self, JsonError> {
        let text = std::str::from_utf8(source).map_err(|error| JsonError {
            offset: error.valid_up_to(),
            message: "JSON input is not valid UTF-8".to_owned(),
        })?;
        Self::parse(text)
    }

    /// Parse product JSON under a caller-supplied non-widening envelope.
    ///
    /// # Errors
    /// Returns the first syntax, type, duplicate-key, integer, or resource
    /// violation with its byte offset.
    pub fn parse_with_limits(source: &str, limits: JsonLimits) -> Result<Self, JsonError> {
        if source.len() > limits.max_bytes {
            return Err(JsonError {
                offset: 0,
                message: "Incomplete.Resource_limit: JSON byte budget exceeded".to_owned(),
            });
        }
        let mut parser = Parser {
            source: source.as_bytes(),
            offset: 0,
            limits,
            values: 0,
        };
        let value = parser.value(0)?;
        parser.whitespace();
        if parser.offset != source.len() {
            return Err(parser.error("trailing JSON input"));
        }
        Ok(value)
    }

    #[must_use]
    pub fn canonical(&self) -> String {
        let mut output = String::new();
        self.write_canonical(&mut output);
        output
    }

    #[must_use]
    pub fn canonical_line(&self) -> String {
        let mut output = self.canonical();
        output.push('\n');
        output
    }

    fn write_canonical(&self, output: &mut String) {
        match self {
            Self::Null => output.push_str("null"),
            Self::Boolean(true) => output.push_str("true"),
            Self::Boolean(false) => output.push_str("false"),
            Self::Integer(value) => output.push_str(&value.to_string()),
            Self::String(value) => write_string(output, value),
            Self::Array(values) => {
                output.push('[');
                for (index, value) in values.iter().enumerate() {
                    if index != 0 {
                        output.push(',');
                    }
                    value.write_canonical(output);
                }
                output.push(']');
            }
            Self::Object(fields) => {
                output.push('{');
                for (index, (name, value)) in fields.iter().enumerate() {
                    if index != 0 {
                        output.push(',');
                    }
                    write_string(output, name);
                    output.push(':');
                    value.write_canonical(output);
                }
                output.push('}');
            }
        }
    }

    /// Check that a value is an object containing no unknown field.
    ///
    /// # Errors
    /// Returns a contextual error for a non-object or the first unknown field.
    pub fn exact_object<'a>(
        &'a self,
        context: &str,
        allowed: &[&str],
    ) -> Result<&'a BTreeMap<String, Self>, String> {
        let Self::Object(fields) = self else {
            return Err(format!("{context} must be an object"));
        };
        let allowed: BTreeSet<&str> = allowed.iter().copied().collect();
        if let Some(name) = fields.keys().find(|name| !allowed.contains(name.as_str())) {
            return Err(format!("{context} has unknown field {name}"));
        }
        Ok(fields)
    }

    #[must_use]
    pub fn member(&self, name: &str) -> Option<&Self> {
        match self {
            Self::Object(fields) => fields.get(name),
            _ => None,
        }
    }

    #[must_use]
    pub fn as_str(&self) -> Option<&str> {
        match self {
            Self::String(value) => Some(value),
            _ => None,
        }
    }

    #[must_use]
    pub fn as_i64(&self) -> Option<i64> {
        match self {
            Self::Integer(value) => Some(*value),
            _ => None,
        }
    }

    #[must_use]
    pub fn as_bool(&self) -> Option<bool> {
        match self {
            Self::Boolean(value) => Some(*value),
            _ => None,
        }
    }

    #[must_use]
    pub fn as_array(&self) -> Option<&[Self]> {
        match self {
            Self::Array(values) => Some(values),
            _ => None,
        }
    }
}

fn write_string(output: &mut String, value: &str) {
    output.push('"');
    for character in value.chars() {
        match character {
            '"' => output.push_str("\\\""),
            '\\' => output.push_str("\\\\"),
            '\u{08}' => output.push_str("\\b"),
            '\u{0c}' => output.push_str("\\f"),
            '\n' => output.push_str("\\n"),
            '\r' => output.push_str("\\r"),
            '\t' => output.push_str("\\t"),
            control if control <= '\u{1f}' => {
                use std::fmt::Write as _;
                let _ = write!(output, "\\u{:04x}", u32::from(control));
            }
            other => output.push(other),
        }
    }
    output.push('"');
}

struct Parser<'a> {
    source: &'a [u8],
    offset: usize,
    limits: JsonLimits,
    values: usize,
}

impl Parser<'_> {
    fn error(&self, message: impl Into<String>) -> JsonError {
        JsonError {
            offset: self.offset,
            message: message.into(),
        }
    }

    fn peek(&self) -> Option<u8> {
        self.source.get(self.offset).copied()
    }

    fn take(&mut self) -> Result<u8, JsonError> {
        let Some(value) = self.peek() else {
            return Err(self.error("unexpected end of JSON"));
        };
        self.offset += 1;
        Ok(value)
    }

    fn whitespace(&mut self) {
        while self
            .peek()
            .is_some_and(|byte| matches!(byte, b' ' | b'\t' | b'\r' | b'\n'))
        {
            self.offset += 1;
        }
    }

    fn value(&mut self, depth: u32) -> Result<JsonValue, JsonError> {
        if depth > self.limits.max_depth {
            return Err(self.error("Incomplete.Resource_limit: JSON nesting budget exceeded"));
        }
        self.values = self.values.saturating_add(1);
        if self.values > self.limits.max_values {
            return Err(self.error("Incomplete.Resource_limit: JSON value budget exceeded"));
        }
        self.whitespace();
        match self.peek() {
            Some(b'n') => self.literal(b"null", JsonValue::Null),
            Some(b't') => self.literal(b"true", JsonValue::Boolean(true)),
            Some(b'f') => self.literal(b"false", JsonValue::Boolean(false)),
            Some(b'"') => self.string().map(JsonValue::String),
            Some(b'[') => self.array(depth + 1),
            Some(b'{') => self.object(depth + 1),
            Some(b'-' | b'0'..=b'9') => self.number(),
            Some(_) | None => Err(self.error("expected a JSON value")),
        }
    }

    fn literal(&mut self, literal: &[u8], value: JsonValue) -> Result<JsonValue, JsonError> {
        for expected in literal {
            if self.take()? != *expected {
                return Err(self.error(format!("expected {}", String::from_utf8_lossy(literal))));
            }
        }
        Ok(value)
    }

    fn string(&mut self) -> Result<String, JsonError> {
        if self.take()? != b'"' {
            return Err(self.error("expected string"));
        }
        let mut output = Vec::new();
        loop {
            match self.take()? {
                b'"' => {
                    return String::from_utf8(output)
                        .map_err(|_| self.error("decoded JSON string is not valid UTF-8"));
                }
                b'\\' => match self.take()? {
                    b'"' => output.push(b'"'),
                    b'\\' => output.push(b'\\'),
                    b'/' => output.push(b'/'),
                    b'b' => output.push(0x08),
                    b'f' => output.push(0x0c),
                    b'n' => output.push(b'\n'),
                    b'r' => output.push(b'\r'),
                    b't' => output.push(b'\t'),
                    b'u' => {
                        let first = self.unicode_escape()?;
                        let scalar = if (0xd800..=0xdbff).contains(&first) {
                            if self.take()? != b'\\' || self.take()? != b'u' {
                                return Err(self
                                    .error("high surrogate must be followed by a low surrogate"));
                            }
                            let second = self.unicode_escape()?;
                            if !(0xdc00..=0xdfff).contains(&second) {
                                return Err(self
                                    .error("high surrogate must be followed by a low surrogate"));
                            }
                            0x1_0000 + ((first - 0xd800) << 10) + (second - 0xdc00)
                        } else if (0xdc00..=0xdfff).contains(&first) {
                            return Err(self.error("lone low surrogate is invalid"));
                        } else {
                            first
                        };
                        let Some(character) = char::from_u32(scalar) else {
                            return Err(self.error("invalid Unicode scalar"));
                        };
                        let mut encoded = [0u8; 4];
                        output.extend_from_slice(character.encode_utf8(&mut encoded).as_bytes());
                    }
                    _ => return Err(self.error("invalid string escape")),
                },
                control if control < 0x20 => {
                    return Err(self.error("unescaped control character"));
                }
                byte => output.push(byte),
            }
        }
    }

    fn unicode_escape(&mut self) -> Result<u32, JsonError> {
        let mut value = 0u32;
        for _ in 0..4 {
            let digit = match self.take()? {
                byte @ b'0'..=b'9' => u32::from(byte - b'0'),
                byte @ b'a'..=b'f' => u32::from(byte - b'a' + 10),
                byte @ b'A'..=b'F' => u32::from(byte - b'A' + 10),
                _ => return Err(self.error("invalid Unicode escape")),
            };
            value = (value << 4) | digit;
        }
        Ok(value)
    }

    fn number(&mut self) -> Result<JsonValue, JsonError> {
        let start = self.offset;
        if self.peek() == Some(b'-') {
            self.offset += 1;
        }
        match self.peek() {
            Some(b'0') => {
                self.offset += 1;
                if self.peek().is_some_and(|byte| byte.is_ascii_digit()) {
                    return Err(self.error("leading zero in JSON number"));
                }
            }
            Some(b'1'..=b'9') => {
                while self.peek().is_some_and(|byte| byte.is_ascii_digit()) {
                    self.offset += 1;
                }
            }
            _ => return Err(self.error("invalid number")),
        }
        if self
            .peek()
            .is_some_and(|byte| matches!(byte, b'.' | b'e' | b'E'))
        {
            return Err(self.error("canonical product JSON permits integers only"));
        }
        let raw = std::str::from_utf8(&self.source[start..self.offset])
            .map_err(|_| self.error("invalid integer"))?;
        let number = raw
            .parse::<i64>()
            .map_err(|_| self.error("integer is out of range"))?;
        Ok(JsonValue::Integer(number))
    }

    fn array(&mut self, depth: u32) -> Result<JsonValue, JsonError> {
        let _ = self.take()?;
        self.whitespace();
        if self.peek() == Some(b']') {
            self.offset += 1;
            return Ok(JsonValue::Array(Vec::new()));
        }
        let mut values = Vec::new();
        loop {
            values.push(self.value(depth)?);
            self.whitespace();
            match self.take()? {
                b']' => return Ok(JsonValue::Array(values)),
                b',' => {}
                _ => return Err(self.error("expected ',' or ']'")),
            }
        }
    }

    fn object(&mut self, depth: u32) -> Result<JsonValue, JsonError> {
        let _ = self.take()?;
        self.whitespace();
        if self.peek() == Some(b'}') {
            self.offset += 1;
            return Ok(JsonValue::Object(BTreeMap::new()));
        }
        let mut fields = BTreeMap::new();
        loop {
            self.whitespace();
            if self.peek() != Some(b'"') {
                return Err(self.error("JSON object key must be a string"));
            }
            let name = self.string()?;
            if fields.contains_key(&name) {
                return Err(self.error(format!("duplicate JSON object key: {name}")));
            }
            self.whitespace();
            if self.take()? != b':' {
                return Err(self.error("expected ':'"));
            }
            let value = self.value(depth)?;
            fields.insert(name, value);
            self.whitespace();
            match self.take()? {
                b'}' => return Ok(JsonValue::Object(fields)),
                b',' => {}
                _ => return Err(self.error("expected ',' or '}'")),
            }
        }
    }
}

impl From<&str> for JsonValue {
    fn from(value: &str) -> Self {
        Self::String(value.to_owned())
    }
}

impl From<String> for JsonValue {
    fn from(value: String) -> Self {
        Self::String(value)
    }
}

impl From<bool> for JsonValue {
    fn from(value: bool) -> Self {
        Self::Boolean(value)
    }
}

impl From<i64> for JsonValue {
    fn from(value: i64) -> Self {
        Self::Integer(value)
    }
}

#[cfg(test)]
mod tests {
    use super::JsonValue;

    #[test]
    fn canonical_contract() {
        let value =
            JsonValue::parse(" { \"z\" : [true, null], \"a\":\"é\\n\" } ").expect("valid fixture");
        assert_eq!(value.canonical(), "{\"a\":\"é\\n\",\"z\":[true,null]}");
    }

    #[test]
    fn strict_input_contract() {
        for invalid in [
            "{\"a\":1,\"a\":2}",
            "01",
            "1.0",
            "9223372036854775808",
            "\"\\ud800\"",
            "[1,]",
        ] {
            assert!(JsonValue::parse(invalid).is_err(), "accepted {invalid}");
        }
    }

    #[test]
    fn surrogate_pair_decodes() {
        assert_eq!(
            JsonValue::parse("\"\\ud83d\\ude00\"").expect("valid fixture"),
            JsonValue::String("😀".to_owned())
        );
    }
}
